PR: Add cookiecutter widget

goanpeca opened the pull request:
<!--- Make sure to read the Contributing Guidelines:                   --->
<!--- https://github.com/spyder-ide/spyder/blob/master/CONTRIBUTING.md --->
<!--- and follow PEP 8, PEP 257 and Spyder's code style:               --->
<!--- https://github.com/spyder-ide/spyder/wiki/Dev:-Coding-Style      --->

## Description of Changes

* [x] Wrote at least one-line docstrings (for any new functions)
* [x] Added unit test(s) covering the changes (if testable)
<!--- Remember that an image/animation is worth a thousand words! --->
* [x] Included a screenshot or animation (if affecting the UI, see [Licecap](https://www.cockos.com/licecap/))


<!--- Explain what you've done and why --->

Add a generic cookiecutter parser so we can generate a UI on the fly with the correct widgets. This work will be used to enable cookicutter projects of different types:
- Data science
- Python Package
- Spyder Plugin. (Example in image)


<img width="564" alt="Screen Shot 2020-06-02 at 14 46 00" src="https://user-images.githubusercontent.com/3627835/83562799-c8b21e00-a4df-11ea-8970-5310cef6b185.png">


### Issue(s) Resolved

<!--- List the issue(s) below, in the form "Fixes #1234"; one per line --->

Fixes #12938


### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

<!--- TYPE YOUR USER/NAME AFTER THE FOLLOWING: --->
I certify the above statement is true and correct: @goanpeca 

<!--- Thanks for your help making Spyder better for everyone! --->

pep8speaks commented:
Hello @goanpeca! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`spyder/plugins/projects/utils/qcookiecutter.py`](https://github.com/spyder-ide/spyder/blob/40a0795e20c9f46ef0cefd8efd020f978a31c357/spyder/plugins/projects/utils/qcookiecutter.py):

> [Line 19:34](https://github.com/spyder-ide/spyder/blob/40a0795e20c9f46ef0cefd8efd020f978a31c357/spyder/plugins/projects/utils/qcookiecutter.py#L19): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace

* In the file [`spyder/plugins/projects/widgets/qcookiecutter.py`](https://github.com/spyder-ide/spyder/blob/40a0795e20c9f46ef0cefd8efd020f978a31c357/spyder/plugins/projects/widgets/qcookiecutter.py):

> [Line 143:1](https://github.com/spyder-ide/spyder/blob/40a0795e20c9f46ef0cefd8efd020f978a31c357/spyder/plugins/projects/widgets/qcookiecutter.py#L143): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 232:50](https://github.com/spyder-ide/spyder/blob/40a0795e20c9f46ef0cefd8efd020f978a31c357/spyder/plugins/projects/widgets/qcookiecutter.py#L232): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 298:1](https://github.com/spyder-ide/spyder/blob/40a0795e20c9f46ef0cefd8efd020f978a31c357/spyder/plugins/projects/widgets/qcookiecutter.py#L298): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 301:1](https://github.com/spyder-ide/spyder/blob/40a0795e20c9f46ef0cefd8efd020f978a31c357/spyder/plugins/projects/widgets/qcookiecutter.py#L301): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace

* In the file [`spyder/plugins/projects/widgets/tests/test_qcookiecutter.py`](https://github.com/spyder-ide/spyder/blob/40a0795e20c9f46ef0cefd8efd020f978a31c357/spyder/plugins/projects/widgets/tests/test_qcookiecutter.py):

> [Line 29:1](https://github.com/spyder-ide/spyder/blob/40a0795e20c9f46ef0cefd8efd020f978a31c357/spyder/plugins/projects/widgets/tests/test_qcookiecutter.py#L29): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1

ccordoba12 commented on .github/scripts/install.sh:
```diff
@@ -56,6 +56,9 @@ popd
 # To check our manifest
 pip install check-manifest
 
+# TODO: update setup.py when package is avalable in defaults
```
For alphas and betas the idea is to add new packages to our channel and install them from there (else the first conda package we create for 5.0 wouldn't have the right dependencies).

So please do that and change the installation script and dependencies files below accordingly.

goanpeca replied:
If I try to add the `spyder`channel to the installation of `conda.txt` it will not work and enter in conflicts. So the spyder channel needs to be emptyfrom anything other than the extra dependencies not found in defaults.

goanpeca replied:
Ok fixed

ccordoba12 replied:
Please remove this comment and the line below then.

ccordoba12 commented on spyder/plugins/projects/widgets/qcookiecutter.py:
```diff
@@ -0,0 +1,419 @@
+# -*- coding: utf-8 -*-
```
I don't like the name of this file too much. What about simply naming it `cookiecutter_qt.py`?

ccordoba12 replied:
Also, please change the line endings of this file to LF.

goanpeca replied:
I really prefer `qcookiecutter` to `cookiecutter_qt`

goanpeca replied:
Not going to change.

ccordoba12 replied:
Ok, let's leave it then. However, you haven't changed the line endings of this file to LF, as I mentioned above.

ccordoba12 commented on spyder/plugins/projects/widgets/qcookiecutter.py:
```diff
@@ -0,0 +1,419 @@
+# -----------------------------------------------------------------------------
+"""
+Cookiecutter widget.
+"""
```
Please add a blank line below this one.

goanpeca replied:
Ok

goanpeca replied:
Fixed

ccordoba12 commented on spyder/plugins/projects/widgets/tests/test_qcookiecutter.py:
```diff
@@ -0,0 +1,152 @@
+#
+"""
+Tests for qcookiecutter widget.
+"""
```
```suggestion
"""

```

goanpeca replied:
Fixed

ccordoba12 commented on spyder/plugins/projects/widgets/tests/test_qcookiecutter.py:
```diff
@@ -0,0 +1,152 @@
+# -*- coding: utf-8 -*-
```
If you decide to change the name of `qcookiecutter.py`, please don't forget to rename this one too.

goanpeca replied:
No change

ccordoba12 commented on spyder/plugins/projects/utils/cookie.py:
```diff
@@ -0,0 +1,55 @@
+#
+# Copyright © Spyder Project Contributors
+# Licensed under the terms of the MIT License
+# (see spyder/__init__.py for details)
```
```suggestion
# (see spyder/__init__.py for details)

```

goanpeca replied:
Fixed

ccordoba12 commented on spyder/plugins/projects/widgets/qcookiecutter.py:
```diff
@@ -0,0 +1,420 @@
+# Licensed under the terms of the MIT License
+# (see spyder/__init__.py for details)
+# -----------------------------------------------------------------------------
+"""
```
Please add a blank line above this line.

goanpeca replied:
Fixed

goanpeca commented:
>You left the line that installs cookiecutter from pip, so we can't be sure that the conda installation is working as expected.

Fixed

>Besides, please change the line endings of qcookiecutter.py to LF.

Fixed @ccordoba12
